src/hotspot/share/gc/shenandoah/shenandoahMark.inline.hpp

```diff
@@ -54,6 +68,7 @@ void ShenandoahMark::do_task(ShenandoahObjToScanQueue* q, T* cl, ShenandoahLiveD
 
   if (task->is_not_chunked()) {
     if (obj->is_instance()) {
+      dedup_string<STRING_DEDUP>(obj, req);
```

Comment:
I do wonder if it is faster to _first_ let the closure process the outbound references, and then do dedup for the object? This way, parallel GC workers can do work while we are calling `dedup_string` here. This is one of the reasons why old code did dedup check after the queue push. In other words, move this call at the end of this block.